hibernate-core/src/main/java/org/hibernate/query/sqm/tree/predicate/SqmMemberOfPredicate.java

```diff
@@ -6,37 +6,41 @@
  */
 package org.hibernate.query.sqm.tree.predicate;
 
+import org.hibernate.metamodel.model.domain.SimpleDomainType;
 import org.hibernate.query.sqm.NodeBuilder;
 import org.hibernate.query.sqm.SemanticQueryWalker;
 import org.hibernate.query.sqm.tree.SqmCopyContext;
-import org.hibernate.query.sqm.tree.domain.SqmPath;
 import org.hibernate.query.sqm.tree.domain.SqmPluralValuedSimplePath;
 import org.hibernate.query.sqm.tree.expression.SqmExpression;
 
+import static org.hibernate.query.sqm.internal.TypecheckUtil.areTypesComparable;
+
 /**
  * @author Steve Ebersole
  */
 public class SqmMemberOfPredicate extends AbstractNegatableSqmPredicate {
 	private final SqmExpression<?> leftHandExpression;
-	private final SqmPath<?> pluralPath;
+	private final SqmPluralValuedSimplePath<?> pluralPath;
 
-	public SqmMemberOfPredicate(SqmExpression<?> leftHandExpression, SqmPath<?> pluralPath, NodeBuilder nodeBuilder) {
+	public SqmMemberOfPredicate(SqmExpression<?> leftHandExpression, SqmPluralValuedSimplePath<?> pluralPath, NodeBuilder nodeBuilder) {
 		this( leftHandExpression, pluralPath, false, nodeBuilder );
 	}
 
 	public SqmMemberOfPredicate(
 			SqmExpression<?> leftHandExpression,
-			SqmPath<?> pluralPath,
+			SqmPluralValuedSimplePath<?> pluralPath,
 			boolean negated,
 			NodeBuilder nodeBuilder) {
 		super( negated, nodeBuilder );
 
 		this.pluralPath = pluralPath;
 		this.leftHandExpression = leftHandExpression;
 
-		leftHandExpression.applyInferableType(
-				( (SqmPluralValuedSimplePath<?>) pluralPath ).getReferencedPathSource().getElementType()
-		);
+		final SimpleDomainType<?> simpleDomainType = pluralPath.getReferencedPathSource().getElementType();
+
+		areTypesComparable( leftHandExpression.getNodeType(), simpleDomainType, nodeBuilder.getSessionFactory() );
```

Comment:
This method just returns a boolean. You still have to check the result and throw an exception if the result is `false`.

Reply:
:face_with_head_bandage: